src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -582,22 +591,22 @@ private bool CheckIfPathIsNested(ReadOnlySpan<char> eventPath)
                 return _includeChildren || _fullDirectory.AsSpan().StartsWith(System.IO.Path.GetDirectoryName(eventPath), StringComparison.OrdinalIgnoreCase);
             }
 
-            private unsafe int FindRenameChangePairedChange(
+            private unsafe int? FindRenameChangePairedChange(
                 int currentIndex,
-                FSEventStreamEventFlags* eventFlags,
-                int numEvents)
+                Span<FSEventStreamEventFlags> flags, Span<FSEventStreamEventId> ids)
             {
-                // Start at one past the current index and try to find the next Rename item, which should be the old path.
-                for (int i = currentIndex + 1; i < numEvents; i++)
+                int nextIndex = currentIndex + 1;
+
+                if (nextIndex >= flags.Length)
+                    return null;
+
+                if (ids[currentIndex] + 1 == ids[nextIndex] &&
+                    IsFlagSet(flags[nextIndex], FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed))
                 {
-                    if (IsFlagSet(eventFlags[i], FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed))
-                    {
-                        // We found match, stop looking
-                        return i;
-                    }
+                    return nextIndex;
                 }
 
-                return int.MinValue;
+                return null;
             }
 
             private static bool IsFlagSet(FSEventStreamEventFlags flags, FSEventStreamEventFlags value)
```

Comment:
I know it's not part of your changeset but we should remove this method and use `HasFlag`, which is now properly optimized.